### ISSUE TITLE
different workdir for native groovy code in process

### ISSUE THREAD

**Issue by monfortm** [nextflow-io/nextflow#942]
Hi guys,
Great work with Nextflow, it's a very nice piece of software!
I'm here to report a bug, but this might be me missing out something from the doc or else.
I'm starting to play around with Nextflow so if this is the expected behavior I would really appreciate to be directed to the piece of doc I missed/misunderstood.

I'll provide more info as needed

Cheers,
M

## Bug report 

When I use file.mklink('relative/filepath') in an exec codeblock, the file is linked to basedir/relative/filepath instead of workdir/[hash]/[hash]/relative/filepath

### Expected behavior and actual behavior

When I use groovy code in an "exec:" block, the working directory is the workflow directory, but when I use e.g. shell code, the working directory is - as expected - work/[hash]/[hash]/.
Thus with file.mklink() it is not possible to capture output file with the ouptut directive syntax AND therefore the publishDir functionality does not work.

### Program output 

[see the actual code below]

WITH BASH CODE:

```
N E X T F L O W  ~  version 18.10.1
Launching `test-links.nf` [scruffy_wright] - revision: 805b07ec1d
[warm up] executor > local
[f1/1332c5] Submitted process > linkSh (1)
ln -s [abspath]/test-links/in-file1.txt ./sh-file-link.txt
```

WITH NATIVE:

```
N E X T F L O W  ~  version 18.10.1
Launching `test-links.nf` [special_woese] - revision: 07f58d1398
[warm up] executor > local
[06/0dbf1a] Submitted process > linkExec (1)
ERROR ~ Error executing process > 'linkExec (1)'

Caused by:
  ./exec-file-link.txt

Source block:
  f = file(infile)
  f.mklink('./exec-file-link.txt')

Work dir:
  [abspath]/test-links/work/06/0dbf1a7e52984e386501ccb3fe7d78

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`

 -- Check '.nextflow.log' file for details
```

### Steps to reproduce the problem

This is the simplest piece of code I could come up with, the buggy part is commented out:

test.nf
```
pDirExec = "exec-dir"
pDirSh = "sh-dir"

sh_ch = Channel.from("$baseDir/in-file1.txt")
exec_ch = Channel.from("$baseDir/in-file2.txt")

process linkSh {
  publishDir path: "${pDirSh}"

  input:
  val f from sh_ch 

  output:
  file 'sh-file-link.txt' into sh_out

  script:
  """
  echo "ln -s ${f} ./sh-file-link.txt"
  ln -s ${f} ./sh-file-link.txt
  """
}

// process linkExec {
//   publishDir path: "${pDirExec}"

//   input:
//   val infile from exec_ch

//   output:
//   file 'exec-file-link.txt' into exec_out

//   exec:
//   f = file(infile)
//   f.mklink('./exec-file-link.txt')
// }
```

### Environment 

* Nextflow version:
N E X T F L O W
version 18.10.1 build 5003
last modified 24-10-2018 14:03 UTC (16:03 CEST)
* Java version: 
openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)
* Operating system:
Linux 3.10.0-862.3.2.el7.x86_64 GNU/Linux

**Comment by pditommaso** [comment]
Yes, file handling when using `exec:` is quirky and needs to be improved. Currently it's expected that files operations are resolved against the work directory given by the `task.workDir` implicit variable. 

Related to #378 #9

**Comment by pachiras** [comment]
I agree with @darnells.
I am always returning to this page every time when I use `ecec:` section and remembering that `task.workDir` must be used for using files.

**Comment by pachiras** [comment]
Thanks for your code update. Right, we can write a 'workable' code with `task.workDir`. But it would be nice if we could write the same code without even thinking about it.

**Comment by stevekm** [comment]
Are there any plans to revisit this?

Would be really, really great if we could use Groovy code inside processes seamlessly the same way that `script` is used. 

I would really love to be able to get rid of my collections of misc. bash, python, etc., scripts to do menial file parsing tasks, and just do them directly in Groovy with `exec` instead without any quirky "gotchas"

**Comment by VasLem** [comment]
+1

**Comment by bentsherman** [comment]
Came across this issue today while testing some `exec` processes. Wanted to share some notes on how I see this issue.

- native tasks don't have a script/subprocess, so there is no need for `env`, `stdin`, `stdout`

- there is a task directory, so `path` output still makes sense as a way to collect output files, though you could also just output a variable pointing to the files

- `path` input doesn't work at all because native tasks don't stage input files. Also there is a [bug](https://github.com/nextflow-io/nextflow/blob/82de4bfe726da274999cb6a5e666320df2a6f18d/modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy#L1950-L1953) which ensures that `path` inputs cannot be accessed in the `exec` block. I think I can fix this

- new files must be explicitly resolved against `task.workDir` when they are created, this is well known. We could override the `file` method in the task body to automatically do this when creating a new file, I think that would cover most use cases. But if the task calls a function which creates files under the hood, that might be harder to intercept

- `val` inputs and outputs work just fine, but there is some missing functionality under the hood around provenance tracking and automatic cleanup. Should be addressed by #4553 which will stage files based on an actual type declaration rather than the `path` qualifier

tl;dr -- just use `val` for file inputs and resolve output files against `task.workDir`

@VasLem if you could share a code example of what you are trying to do, I should be able to help you make it work

**Comment by VasLem** [comment]
Thank you for coming back to address this issue and provide your insights @bentsherman ! Certainly, an example is to scan a directory and extract the matching files, returning them, or their paths, to the next process, like this:
```
process readFromDirectoryExec {
    input:
        val sample_dir
        val sample_name
    output:
        tuple val(sample_name), val(output)

    exec:
        output = file(sample_dir).listFiles().filter{ it -> (it.getName() =~ /${sample_name}.*\.f(?:ast)?[qa](?:\.gz)?/ )}
        def bucket = (sample_dir =~ /gs\:\/\/([\w-]+)/[0]​[1])
        output = output.collect{ "gs://${bucket}/${it}" }
}
```
When running this example, I noticed I was getting a warning that the local executor was used instead (I am not sure this is part of the latest release, I am using the edge one, for some days now). The process completed successfully. I noticed that the returned files did not contain the gs bucket, so I had to manually prepend it. However, apart from this minor requirement, (I am not sure if the following needs to get its own issue) then I tried to connect another process, that expects an input tuple (val, path), like the following:
```
process headFiles {
    input:
        tuple val(sample_name), path(files, arity='1..*')
    shell:
        '''
        for f in !{files}
        do
            echo "Head of file $f:"
            head -n 10 $f
        done 
        '''
}
```

and there I got an error:
`ERROR ~ Could not find matching constructor for: nextflow.script.TokenPathCall(nextflow.script.TokenVar, String)`

The error occurred even before running the pipeline. Which means that the bug with the `path` not being available to the exec process makes it not possible to supply paths as inputs to a next process. For clarification, here is the workflow I used:

workflow {
    out = readFromDirectoryExec("gs://private_bucket/emseq/downsampled_0.05/CRC_DE02NGSLABD101656", "CRC_DE02NGSLABD101656")
    headFiles(out)
}

**Comment by bentsherman** [comment]
@VasLem so you want `readFromDirectoryExec` to scan a directory of files in a GCS bucket and filter some matching files

since you are not modifying the file contents or writing new files, why don't you just do this in a `map` operator?

Something like this:
```groovy
ch_sample | map { sample_dir, sample_name ->
  output = file(sample_dir)
    .listFiles()
    .filter { it -> (it.getName() =~ /${sample_name}.*\.f(?:ast)?[qa](?:\.gz)?/ ) }

  return [ sample_name, output ]
}
```

**Comment by VasLem** [comment]
Hi @bentsherman I know that there is a way to use workflow code to perform the exact same thing. The reason why I want it to be a process is so that it is modular, cached and it is listed as a step in the Nextflow report. Any commands that are within a workflow are transparent and not easy to isolate in case of a bug. Maintaining a code-base used by many users who don't all have groovy or deep programming knowledge renders having such processes as an easy way to demystify the code and simplify their life. The first time I had done that was also because I was under the impression there was no boilerplate supporting directory wildcard resolution in the body of the workflow.

**Comment by bentsherman** [comment]
Fair enough, it is easily refactored into a process:
```groovy
process readFromDirectoryExec {
  input:
  tuple val(sample_dir), val(sample_name)

  output:
  tuple val(sample_name), val(output)

  exec:
  output = file(sample_dir)
    .listFiles()
    .filter { it -> (it.getName() =~ /${sample_name}.*\.f(?:ast)?[qa](?:\.gz)?/ ) }
}
```

**Comment by bentsherman** [comment]
The point is that you don't need to do anything with the path names. They are already full GCS paths and they will be staged into downstream tasks as usual
